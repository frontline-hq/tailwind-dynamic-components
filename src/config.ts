import path from "node:path";
import { readFile, writeFile, stat } from "node:fs/promises";
import { pathToFileURL } from "node:url";
import type { Config as SvelteConfig } from "@sveltejs/kit";
import { findDepPkgJsonPath } from "vitefu";
import { libraryName } from "./library.config";
import { CompoundStyles, Styles } from "./register";

const configFileName = `${libraryName}.config.js`;

const cwdFolderPath = process.cwd();
const libraryConfigFilePath = path.resolve(cwdFolderPath, configFileName);

export const doesPathExist = async (path: string) =>
    !!(await stat(path).catch(() => false));

type VersionString = `${number}.${number}.${number}`;

type LibraryConfigModule = {
    defineConfig: DefineConfig;
};

type LibraryConfig = {
    debug: boolean;
    registrations: Array<CompoundStyles | Styles>;
};

type DefineConfig = () => LibraryConfig;

export type TransformConfig = {
    cwdFolderPath: string;
    debug?: boolean;
    usesTypeScript: boolean;
    library: LibraryConfig;
    svelteKit?: {
        version: VersionString | undefined;
        rootRoutesFolder: string;
        files: {
            appTemplate: string;
            routes: string;
            serverHooks: string;
        };
    };
};

export const getTransformConfig = async (): Promise<TransformConfig> => {
    // eslint-disable-next-line no-async-promise-executor
    return new Promise<TransformConfig>(async resolve => {
        await createLibraryConfigIfNotPresentYet();

        const libraryConfigModule = (await import(
            pathToFileURL(libraryConfigFilePath).toString()
        )) as LibraryConfigModule;
        const libraryConfig = await libraryConfigModule.defineConfig();

        // Svelte specific data

        let svelteKitConfigData: TransformConfig["svelteKit"];

        try {
            const { default: svelteConfig } = (await import(
                pathToFileURL(
                    path.resolve(cwdFolderPath, "svelte.config.js")
                ).toString()
            )) as { default: SvelteConfig };
            const files = {
                appTemplate: path.resolve(
                    cwdFolderPath,
                    svelteConfig.kit?.files?.appTemplate ||
                        path.resolve("src", "app.html")
                ),
                routes: path.resolve(
                    cwdFolderPath,
                    svelteConfig.kit?.files?.routes ||
                        path.resolve("src", "routes")
                ),
                serverHooks: path.resolve(
                    cwdFolderPath,
                    svelteConfig.kit?.files?.hooks?.server ||
                        path.resolve("src", "hooks.server")
                ),
            };

            const rootRoutesFolder = path.resolve(files.routes, "");

            // TODO: find a more reliable way (https://github.com/sveltejs/kit/issues/9937)
            const svelteKitVersion =
                await getInstalledVersionOfPackage("@sveltejs/kit");
            svelteKitConfigData = {
                version: svelteKitVersion,
                files,
                rootRoutesFolder,
            };
        } catch (error) {
            console.log("Svelte detection failed. skipped.");
        }

        const usesTypeScript = await doesPathExist(
            path.resolve(cwdFolderPath, "tsconfig.json")
        );
        resolve({
            cwdFolderPath,
            debug: !!libraryConfig.debug,
            library: libraryConfig,
            usesTypeScript,
            svelteKit: svelteKitConfigData,
        });
    });
};

const createLibraryConfigIfNotPresentYet = async () => {
    const libraryConfigExists = await doesPathExist(libraryConfigFilePath);
    if (libraryConfigExists) return;

    /* TODO: Inject config type definitin into defineConfig file */

    return writeFile(
        libraryConfigFilePath,
        `
export async function defineConfig() {
	return {
		debug: false,
        /* Styles within CompoundStyles have to be registered seperately for their detection */
        registrations: []
	}
}
`
    );
};

const getInstalledVersionOfPackage = async (pkg: string) => {
    const pkgJsonPath = await findDepPkgJsonPath(pkg, cwdFolderPath);
    if (!pkgJsonPath) return undefined;

    const pkgJson = JSON.parse(
        await readFile(pkgJsonPath, { encoding: "utf-8" })
    );
    return pkgJson.version;
};
