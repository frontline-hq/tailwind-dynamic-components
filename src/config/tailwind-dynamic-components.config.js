import { styles1 } from "../register.test";

export async function defineConfig() {
    return {
        debug: true,
        registrations: [styles1],
    };
}
