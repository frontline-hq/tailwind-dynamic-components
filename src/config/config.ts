import path from "path";
import { writeFile, stat /* readFile */ } from "fs/promises";
import { pathToFileURL } from "url";
/* import { VERSION } from "@sveltejs/kit"; */
import type { Config as SvelteConfig } from "@sveltejs/kit";
/* import { findDepPkgJsonPath } from "vitefu"; */
import { libraryName } from "../library.config";
import { CompoundStyles, Styles } from "../register";

export const configFileName = `${libraryName}.config.js` as const;

const cwdFolderPath = process.cwd();
const getLibraryConfigFilePath = async (directory: string = process.cwd()) =>
    path.resolve(await findConfigFileFolder(directory), configFileName);

export async function findConfigFileFolder(directory: string = process.cwd()) {
    if (
        (await doesPathExist(path.join(directory, configFileName))) ||
        directory === process.cwd()
    )
        return directory;
    const parentDirectory = path.join(directory, "..");
    return findConfigFileFolder(parentDirectory);
}

export const doesPathExist = async (path: string) =>
    !!(await stat(path).catch(() => false));

/* type VersionString = `${number}.${number}.${number}`; */

type LibraryConfigModule = {
    defineConfig: DefineConfig;
};

type LibraryConfig = {
    debug: boolean;
    registrations: Array<CompoundStyles | Styles>;
};

type DefineConfig = () => LibraryConfig;

export type TransformConfig = {
    cwdFolderPath: string;
    debug?: boolean;
    usesTypeScript: boolean;
    library: LibraryConfig;
    svelteKit?: {
        version: string;
        rootRoutesFolder: string;
        files: {
            appTemplate: string;
            routes: string;
            serverHooks: string;
        };
    };
};

export function flattenAndCheckRegistrations(
    registrations: Array<CompoundStyles | Styles>
) {
    // 1. Add all subregistrations of CompoundStyles
    const completeRegistrations = registrations.flatMap((r, _, arr) => {
        if (r instanceof CompoundStyles)
            return [r, ...Object.values(r.styles)] as typeof arr;
        return [r];
    });

    // 2. Check that there are no duplicate descriptions.
    const duplicateRegistration = completeRegistrations
        .map(r => r.description)
        .filter((item, index, array) => array.indexOf(item) !== index);
    if (duplicateRegistration.length > 0)
        throw new Error(
            `Found duplicate registration ${String(duplicateRegistration)}.`
        );
    return completeRegistrations;
}

export async function getTransformConfig(
    directory: string = process.cwd()
): Promise<TransformConfig> {
    await createLibraryConfigIfNotPresentYet();

    const libraryConfigModule = (await import(
        pathToFileURL(await getLibraryConfigFilePath(directory)).toString()
    )) as LibraryConfigModule;
    const libraryConfig = await libraryConfigModule.defineConfig();

    // Process registrations in libraryconfig.
    const completeRegistrations = flattenAndCheckRegistrations(
        libraryConfig.registrations
    );
    libraryConfig.registrations = completeRegistrations;
    // Svelte specific data

    let svelteKitConfigData: TransformConfig["svelteKit"];

    try {
        const { default: svelteConfig } = (await import(
            pathToFileURL(
                path.resolve(cwdFolderPath, "svelte.config.js")
            ).toString()
        )) as { default: SvelteConfig };
        const files = {
            appTemplate: path.resolve(
                cwdFolderPath,
                svelteConfig.kit?.files?.appTemplate ||
                    path.resolve("src", "app.html")
            ),
            routes: path.resolve(
                cwdFolderPath,
                svelteConfig.kit?.files?.routes || path.resolve("src", "routes")
            ),
            serverHooks: path.resolve(
                cwdFolderPath,
                svelteConfig.kit?.files?.hooks?.server ||
                    path.resolve("src", "hooks.server")
            ),
        };

        const rootRoutesFolder = path.resolve(files.routes, "");

        // TODO: find a more reliable way (https://github.com/sveltejs/kit/issues/9937)
        const svelteKitVersion = (await import("@sveltejs/kit")).VERSION;
        svelteKitConfigData = {
            version: svelteKitVersion,
            files,
            rootRoutesFolder,
        };
    } catch (error) {
        console.log("Svelte detection failed. skipped.");
    }

    const usesTypeScript = await doesPathExist(
        path.resolve(cwdFolderPath, "tsconfig.json")
    );
    // eslint-disable-next-line no-async-promise-executor
    return {
        cwdFolderPath,
        debug: !!libraryConfig.debug,
        library: libraryConfig,
        usesTypeScript,
        svelteKit: svelteKitConfigData,
    };
}

const createLibraryConfigIfNotPresentYet = async (
    directory: string = process.cwd()
) => {
    const libraryConfigExists = await doesPathExist(
        await getLibraryConfigFilePath(directory)
    );
    if (libraryConfigExists) return;

    /* TODO: Inject config type definitin into defineConfig file */

    return writeFile(
        await getLibraryConfigFilePath(directory),
        `
export async function defineConfig() {
	return {
		debug: false,
        /* Styles within CompoundStyles have to be registered seperately for their detection */
        registrations: []
	}
}
`
    );
};

/* const getInstalledVersionOfPackage = async (pkg: string) => {
    const pkgJsonPath = await findDepPkgJsonPath(pkg, cwdFolderPath);
    if (!pkgJsonPath) return undefined;

    const pkgJson = JSON.parse(
        await readFile(pkgJsonPath, { encoding: "utf-8" })
    );
    return pkgJson.version;
}; */
