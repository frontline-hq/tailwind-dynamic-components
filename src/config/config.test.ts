import { describe, expect, test } from "vitest";
import { TransformConfig, getTransformConfig } from "./config";
import { styles1 } from "../register.test";
import merge from "lodash.merge";
import { Styles } from "../register";

describe("config", () => {
    describe("properties", () => {
        test("debug", async () => {
            expect(
                (
                    merge(await getTransformConfig(__dirname), {
                        library: {
                            debug: false,
                            registrations: [
                                new Styles("b0is", {
                                    dependencies: {
                                        color: ["blue", "gray"],
                                        heights: ["one", "two"],
                                    },
                                }),
                            ],
                        },
                    }) as TransformConfig
                ).debug
            ).toEqual(true);
        });
    });

    describe("getTransformConfig", () => {
        test("cwdFolderPath", async () =>
            expect((await getTransformConfig()).cwdFolderPath).toEqual(
                process.cwd()
            ));
        test("debug", async () =>
            expect((await getTransformConfig(__dirname)).debug).toEqual(true));
        test("library.debug", async () =>
            expect((await getTransformConfig(__dirname)).library.debug).toEqual(
                true
            ));
        test("library.registrations", async () =>
            expect(
                (await getTransformConfig(__dirname)).library.registrations
            ).toEqual([styles1]));
    });
});
