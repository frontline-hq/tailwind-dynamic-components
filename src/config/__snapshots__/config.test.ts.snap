// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`analyze > js > emitted files 1`] = `
Map {
  "FroHnY" => {
    "fileReference": "ref-tdc-FroHnY",
    "styles": Styles {
      "d": [Function],
      "dependencies": {
        "color": [
          "blue",
          "gray",
        ],
        "heights": [
          "one",
          "two",
        ],
      },
      "description": "b0is",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0is\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0is\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "UvbZZuJVSa" => {
    "fileReference": "ref-tdc-UvbZZuJVSa",
    "styles": Styles {
      "d": [Function],
      "dependencies": {
        "color": [
          "blue",
          "gray",
        ],
        "heights": [
          "one",
          "two",
        ],
      },
      "description": "b0is",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0is\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0is\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "DFutYsdTuNsmYdtaFQU" => {
    "fileReference": "ref-tdc-DFutYsdTuNsmYdtaFQU",
    "styles": Styles {
      "d": [Function],
      "dependencies": {
        "color": [
          "blue",
          "gray",
        ],
        "heights": [
          "one",
          "two",
        ],
      },
      "description": "b0is",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0is\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0is\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "FroGui" => {
    "fileReference": "ref-tdc-FroGui",
    "styles": Styles {
      "d": [Function],
      "dependencies": {},
      "description": "b0as",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0as\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0as\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "BFkFG" => {
    "fileReference": "ref-tdc-BFkFG",
    "styles": CompoundStyles {
      "add": [Function],
      "addInline": [Function],
      "compile": [Function],
      "description": "b0s",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0s\\$/g,
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0s\\$\\)/g,
      "propType": {},
      "styles": {
        "b0as": Styles {
          "d": [Function],
          "dependencies": {},
          "description": "b0as",
          "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0as\\$/g,
          "getV": [Function],
          "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
          "literalEncoding": "utf8",
          "matchDescription": [Function],
          "matchModifiers": [Function],
          "matchVariant": [Function],
          "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0as\\$\\)/g,
          "propType": [Function],
          "styles": [
            [],
            [],
          ],
          "variants": [],
          "variantsRegex": /\\^\\$/g,
        },
      },
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "USKtRmcqhVTaNLtGHTvY" => {
    "fileReference": "ref-tdc-USKtRmcqhVTaNLtGHTvY",
    "styles": Styles {
      "d": [Function],
      "dependencies": {},
      "description": "b0as",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0as\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0as\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
}
`;

exports[`analyze > svelte > emitted files 1`] = `
Map {
  "FroHnY" => {
    "fileReference": "ref-tdc-FroHnY",
    "styles": Styles {
      "d": [Function],
      "dependencies": {
        "color": [
          "blue",
          "gray",
        ],
        "heights": [
          "one",
          "two",
        ],
      },
      "description": "b0is",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0is\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0is\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "UvbZZuJVSa" => {
    "fileReference": "ref-tdc-UvbZZuJVSa",
    "styles": Styles {
      "d": [Function],
      "dependencies": {
        "color": [
          "blue",
          "gray",
        ],
        "heights": [
          "one",
          "two",
        ],
      },
      "description": "b0is",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0is\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0is\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "DFutYsdTuNsmYdtaFQU" => {
    "fileReference": "ref-tdc-DFutYsdTuNsmYdtaFQU",
    "styles": Styles {
      "d": [Function],
      "dependencies": {
        "color": [
          "blue",
          "gray",
        ],
        "heights": [
          "one",
          "two",
        ],
      },
      "description": "b0is",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0is\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0is\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "FroGui" => {
    "fileReference": "ref-tdc-FroGui",
    "styles": Styles {
      "d": [Function],
      "dependencies": {},
      "description": "b0as",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0as\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0as\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "BFkFG" => {
    "fileReference": "ref-tdc-BFkFG",
    "styles": CompoundStyles {
      "add": [Function],
      "addInline": [Function],
      "compile": [Function],
      "description": "b0s",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0s\\$/g,
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0s\\$\\)/g,
      "propType": {},
      "styles": {
        "b0as": Styles {
          "d": [Function],
          "dependencies": {},
          "description": "b0as",
          "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0as\\$/g,
          "getV": [Function],
          "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
          "literalEncoding": "utf8",
          "matchDescription": [Function],
          "matchModifiers": [Function],
          "matchVariant": [Function],
          "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0as\\$\\)/g,
          "propType": [Function],
          "styles": [
            [],
            [],
          ],
          "variants": [],
          "variantsRegex": /\\^\\$/g,
        },
      },
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
  "USKtRmcqhVTaNLtGHTvY" => {
    "fileReference": "ref-tdc-USKtRmcqhVTaNLtGHTvY",
    "styles": Styles {
      "d": [Function],
      "dependencies": {},
      "description": "b0as",
      "descriptionRegex": /\\(\\?<=\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\*\\)b0as\\$/g,
      "getV": [Function],
      "identifierAlphabet": "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
      "literalEncoding": "utf8",
      "matchDescription": [Function],
      "matchModifiers": [Function],
      "matchVariant": [Function],
      "modifiersRegex": /\\^\\(\\[\\^\\\\t\\\\n\\\\r\\\\s:\\]\\+:\\)\\+\\(\\?=b0as\\$\\)/g,
      "propType": [Function],
      "styles": [
        [],
        [],
      ],
      "variants": [],
      "variantsRegex": /\\^\\$/g,
    },
  },
}
`;
